File: apps/node/src/generator/vu.ts
import type { Dispatcher } from 'undici';
import type { Scenario } from '@hammr/shared';
import { runIteration } from '../scenario/engine.js';

export type { EventSink } from '../scenario/engine.js';
import type { EventSink } from '../scenario/engine.js';

export interface VUContext {
  vuId: number;
  threadId: number;
  generatorId: string;
  scenario: Scenario;
  // Absolute deadline in performance.now() units. The VU stops when we pass it.
  endAt: number;
}

// A single virtual user. Re-runs the scenario from the top until the thread's
// deadline passes or the abort signal fires. Each iteration starts with a fresh
// variable context (isolation between VUs is guaranteed by this call being
// invoked independently per VU in the caller).
export async function runVU(
  ctx: VUContext,
  agent: Dispatcher,
  sink: EventSink,
  signal: AbortSignal,
): Promise<void> {
  const identity = {
    generatorId: ctx.generatorId,
    threadId: ctx.threadId,
    vuId: ctx.vuId,
  };
  while (!signal.aborted && performance.now() < ctx.endAt) {
    await runIteration(ctx.scenario, agent, sink, signal, identity);
  }
}

File: apps/go-generator/internal/vu/vu.go
// Package vu runs a single virtual user. One goroutine per VU; the goroutine
// re-runs scenario.RunIteration until the run-level context is cancelled.
// Mirrors apps/node/src/generator/vu.ts.
package vu

import (
	"context"
	"math/rand"

	"github.com/Andr0human/hammr-go-generator/internal/scenario"
)

// Context bundles the per-VU identity and scenario. The deadline is carried by
// ctx (use context.WithDeadline / WithTimeout), not a separate endAt field —
// Go idiom diverges from the Node version here.
type Context struct {
	VUID        int
	GeneratorID string
	Scenario    *scenario.Scenario
}

// Run loops scenario iterations until ctx is cancelled or the scenario aborts
// internally. Each iteration starts with a fresh variable map (engine guarantees
// this) so VUs are isolated by construction. Per-VU rng is seeded by the caller
// to keep think-time picks reproducible across ports.
func Run(
	ctx context.Context,
	vc Context,
	doer scenario.HTTPDoer,
	sink scenario.EventSink,
	rng *rand.Rand,
) {
	identity := scenario.VUIdentity{
		GeneratorID: vc.GeneratorID,
		ThreadID:    0, // Go has no thread concept; wire shape kept by emitting 0.
		VUID:        vc.VUID,
	}
	for ctx.Err() == nil {
		// RunIteration is responsible for honouring ctx mid-step. Errors are
		// already surfaced as engine-error events (statusCode=0) inside the
		// engine; we don't propagate them here.
		_ = scenario.RunIteration(ctx, vc.Scenario, doer, sink, identity, rng)
	}
}
